### PR TITLE
Converting Vulkan target to support executable-create2.

### PR DESCRIPTION
**Pull request by benvanik**
This produces a new flatbuffer that supports multiple shared modules per HAL executable, reorganizes per-export information to be per-export, and swaps HAL pipeline layouts with internal Vulkan pipeline layouts that are constructed from metadata in the flatbuffer.

This is a startup time regression for as long as we are not linking modules because descriptor set and pipeline layout reuse now happens per executable and not per VM context. The fix is to link modules together. I am not sure why that was never done :/

Progress on #18154.

### CODE REVIEW

**Comment by benvanik** [comment]
(lit test failures are expected during the transition)

File: runtime/src/iree/hal/drivers/vulkan/pipeline_layout.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_HAL_DRIVERS_VULKAN_NATIVE_PIPELINE_LAYOUT_H_
+#define IREE_HAL_DRIVERS_VULKAN_NATIVE_PIPELINE_LAYOUT_H_
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/vulkan/pipeline_layout.h]: nit: update header guard to match new file name (here and at the bottom of the file)

```suggestion
#ifndef IREE_HAL_DRIVERS_VULKAN_PIPELINE_LAYOUT_H_
#define IREE_HAL_DRIVERS_VULKAN_PIPELINE_LAYOUT_H_
```

File: runtime/src/iree/hal/drivers/vulkan/pipeline_layout.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+// Returns the descriptor set layout with the given |set_index|.
+iree_hal_vulkan_descriptor_set_layout_t* iree_hal_vulkan_pipeline_layout_set(
+    iree_hal_vulkan_pipeline_layout_t* pipeline_layout,
+    iree_host_size_t set_index);
+
+#ifdef __cplusplus
+}  // extern "C"
+#endif  // __cplusplus
+
+#endif  // IREE_HAL_DRIVERS_VULKAN_NATIVE_PIPELINE_LAYOUT_H_
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/vulkan/pipeline_layout.h]:
```suggestion
#endif  // IREE_HAL_DRIVERS_VULKAN_PIPELINE_LAYOUT_H_
```

File: runtime/src/iree/schemas/vulkan_executable_def.fbs [hunk 3]
```diff
@@ -12,42 +12,96 @@ namespace iree.hal.vulkan;
...
+
 // A SPIR-V shader module and runtime pipeline layout description.
 // This information is used to create the VkShaderModule, VkPipelineLayout, and
 // any required VkDescriptorSetLayouts.
 table ExecutableDef {
-  // A map of entry point ordinals to string names as used in the shader module.
-  entry_points:[string];
+  // Exported functions in canonical executable entry point order.
+  // Each creates a single VkPipeline.
+  pipelines:[PipelineDef];
```
Comment (ScottTodd) [inline runtime/src/iree/schemas/vulkan_executable_def.fbs:89]: Are the other fields that refer to the "the exports list" talking about this list of `PipelineDef`s?

BTW, found a good doc: https://docs.vulkan.org/spec/latest/chapters/pipelines.html#pipelines-compute

File: runtime/src/iree/hal/drivers/vulkan/native_executable.cc [hunk 4]
```diff
@@ -27,187 +27,179 @@
...
+        case iree_hal_vulkan_VkDescriptorType_STORAGE_BUFFER:
+          storage_descriptor_count += descriptor_count;
+          break;
+        default:
+          return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                                  "pipeline layout set[%" PRIhsz
+                                  "] binding[%" PRIhsz
+                                  "] has "
+                                  "an unsupported descriptor_type",
+                                  i, j);
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/vulkan/native_executable.cc]: nit: I think these lines can be condensed?
```suggestion
          return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
                                  "pipeline layout set[%" PRIhsz
                                  "] binding[%" PRIhsz
                                  "] has an unsupported descriptor_type",
                                  i, j);
```

File: runtime/src/iree/hal/drivers/vulkan/native_executable.cc [hunk 5]
```diff
@@ -230,88 +222,657 @@ static iree_status_t iree_hal_vulkan_executable_flatbuffer_verify(
...
+//===----------------------------------------------------------------------===//
+// Pipelines
+//===----------------------------------------------------------------------===//
+
+// Creates a pipeline from the set of available pipeline layouts and shader
+// modules and stores it into |out_pipeline|.
+//
+// NOTE: vkCreateComputePipelines takes multiple pipelines but doesn't speed up
+// creation on any known driver; we process one at a time so that we can get
+// better error messages and multithread the pipeline creation ourselves.
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/vulkan/native_executable.cc:661]: Interesting - did you run those speed tests / benchmarks yourself, or did other people report that somewhere? (I'll buy the point about error messages, but startup time is pretty important for us, so would prefer to have high confidence that what we're doing is optimal)

**Review by ScottTodd — APPROVED** [review]

File: runtime/src/iree/schemas/vulkan_executable_def.fbs (hunk 3 above)
Comment (benvanik) [reply runtime/src/iree/schemas/vulkan_executable_def.fbs:89]: good catch

File: runtime/src/iree/hal/drivers/vulkan/native_executable.cc (hunk 5 above)
Comment (benvanik) [reply runtime/src/iree/hal/drivers/vulkan/native_executable.cc:661]: Yeah, drivers are not required and don't multithread pipeline creation internally (as a general rule Vulkan drivers don't maintain their own thread pools in user mode). The only savings is on the call into the driver which is tiny compared to actual creation and shader JIT that usually happens.

From the Vulkan guide:
> For compiling pipelines, vkCreateShaderModule and vkCreateGraphicsPipeline are both allowed to be called from multiple threads at once. A common approach for multithreaded shader compilation is to have a background thread dedicated to it, with it constantly looking into a parallel queue to receive compilation requests, and putting the compiled pipelines into another queue that then the main renderthread will connect to the simulation. This is very important to do if you want to have an engine that doesn’t have a lot of hitching. Compiling shader pipelines can take a very long time, so if you have to compile pipelines at runtime outside of a load screen, then you need to implement such a multithreaded async compile scheme for your game to work well.

**Pull request merged**